lxd/profiles.go

```diff
@@ -43,11 +43,70 @@ var profilesCmd = APIEndpoint{
 var profileCmd = APIEndpoint{
 	Path: "profiles/{name}",
 
-	Delete: APIEndpointAction{Handler: profileDelete, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanDelete, "name")},
-	Get:    APIEndpointAction{Handler: profileGet, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanView, "name")},
-	Patch:  APIEndpointAction{Handler: profilePatch, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
-	Post:   APIEndpointAction{Handler: profilePost, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
-	Put:    APIEndpointAction{Handler: profilePut, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
+	Delete: APIEndpointAction{Handler: profileDelete, AccessHandler: profileAccessHandler(auth.EntitlementCanDelete)},
+	Get:    APIEndpointAction{Handler: profileGet, AccessHandler: profileAccessHandler(auth.EntitlementCanView)},
+	Patch:  APIEndpointAction{Handler: profilePatch, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+	Post:   APIEndpointAction{Handler: profilePost, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+	Put:    APIEndpointAction{Handler: profilePut, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+}
+
+// ctxProfileDetails should be used only for getting/setting profileDetails in the request context.
+const ctxProfileDetails request.CtxKey = "profile-details"
+
+// profileDetails contains fields that are determined prior to the access check. This is set in the request context when
+// addProfileDetailsToRequestContext is called.
+type profileDetails struct {
+	profileName      string
+	effectiveProject api.Project
+}
+
+// addProfileDetailsToRequestContext sets request.CtxEffectiveProjectName (string) and ctxProfileDetails (profileDetails)
+// in the request context.
+func addProfileDetailsToRequestContext(d *Daemon, r *http.Request) error {
+	s := d.State()
+	profileName, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		return err
+	}
+
+	requestProjectName := request.ProjectParam(r)
+	effectiveProject, err := project.ProfileProject(s.DB.Cluster, requestProjectName)
+	if err != nil && api.StatusErrorCheck(err, http.StatusNotFound) {
+		return api.StatusErrorf(http.StatusNotFound, http.StatusText(http.StatusNotFound))
```

Comment:
Why are you masking the error message from project.ProfileProject() with a generic "Not found" message rather than using SmartResponse and which would pass the "profile not found" type error to the caller?

Reply:
> The logic is starting to leak into lots of places around the code base though, I'm starting to think that `response.SmartError` should handle this instead. E.g. if `auth.IsDeniedError` we return a masked error message, otherwise we return the true error.

I think that makes sense.

> We're doing this masking in lots of places now to avoid resource discovery. If the project really doesn't exist, the user will get `Project not found`. If the project does exist but the user does not have access to it, they will get `Not found`.

I do remember discussing that. Although the way you've described it here makes me think, can an end user still figure out if a resource exists by comparing the error strings instead of the response codes?

Reply:
I've updated SmartError to mask not found and forbidden errors to use the default HTTP status text.

The end user can figure out if a resource exists by the error text in some cases I think. I haven't done a full audit of this. It wouldn't work for project names as we don't need to do any logic before the authorization check. The general case is as follows:
1. Caller makes request to view/edit/delete a specific entity.
2. The entity requires some pre-processing before we can perform the access check. This can be for the effective project, or it can be to expand an image or certificate fingerprint.
3. The error returned in the pre-processing step may indicate the resource does not exist with message like "<thing> not found.
4. If the resource is found but the caller does not have permission to view it, the authorizer returns "Not Found".

If the caller gets "Not Found" and not "<thing> not found", they can surmise that the entity does exist. In most cases they already need to be authenticated for this, but some endpoints (images in particular) allow untrusted requests.

I think updating SmartError will catch the majority of cases and I don't think it's a general concern as an attack vector, I'm just being thorough :)